### PR TITLE
ucm2: add profile for the Librem 5

### PR DESCRIPTION
**Pull request by craftyguy**
cc @agx

### CODE REVIEW

**Comment by agx** [comment]
@craftyguy thanks for taking a look at this! Would be great to have it upstream. There's  rework pending here https://source.puri.sm/Librem5/librem5-base/-/merge_requests/311 which might make sense to have upstream right away. What do you think?

**Comment by craftyguy** [comment]
> There's rework pending here https://source.puri.sm/Librem5/librem5-base/-/merge_requests/311 which might make sense to have upstream right away. What do you think?

@agx I saw that, and based on the comments it still seemed WIP with the last update a couple of months ago... So I thought it would be better to upstream a known working config, then it could be tweaked later to perfection.

But obviously if it's in much better shape than what I think from reading the discussion, I'd have no problem updating this patch. My main motivation is to just get *something* upstream to hopefully help with audio on this platform breaking every time alsa is updated.

**Comment by agx** [comment]
Given that reworking the profile seems to have stalled I think this would be good to have as a starting point!

**Comment by agx** [comment]
and just in case:

```
Tested-by: Guido Günther <agx@sigxcpu.org>
```

File: ucm2/NXP/iMX8/Librem_5/HiFi.conf [hunk 1]
```diff
@@ -0,0 +1,180 @@
+SectionVerb {
+	EnableSequence [
+		cset "name='Digital Playback Volume' 100,100"
```
Comment (perexg) [inline ucm2/NXP/iMX8/Librem_5/HiFi.conf:3]: This settings belongs to the BootSection.

File: ucm2/NXP/iMX8/Librem_5/HiFi.conf [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+		cset "name='INPGAL IN2L Switch' off"
+		cset "name='INPGAL IN3L Switch' off"
+		cset "name='INPGAL IN4L Switch' off"
+		cset "name='Input Mixer Switch' off,on"
+		cset "name='SPKOUTL PGA' Mixer"
+		cset "name='SPKOUTR PGA' Mixer"
+	]
+
+	DisableSequence [
+		cset "name='Digital Playback Volume' 0,0"
```
Comment (perexg) [inline ucm2/NXP/iMX8/Librem_5/HiFi.conf:22]: The volumes should not be touched in enable/disable sequences. Keep the user settings.

File: ucm2/NXP/iMX8/Librem_5/HiFi.conf [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+	]
+
+	Value {
```
Comment (perexg) [inline ucm2/NXP/iMX8/Librem_5/HiFi.conf:25]: Remove empty blocks.

File: ucm2/NXP/iMX8/Librem_5/HiFi.conf [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+		"Headphones"
+		"Speaker"
+	]
+
+	EnableSequence [
+		cset "name='Speaker Switch' off,on"
+		cset "name='DAC L/R Swap Switch' on"
+		cset "name='DAC Monomix Switch' on"
+		cset "name='SPKOUTR PGA' DAC"
+		cset "name='Speaker Volume' 121,121"
```
Comment (perexg) [inline ucm2/NXP/iMX8/Librem_5/HiFi.conf:42]: Do not set volume in enable/disable sequences.

File: ucm2/NXP/iMX8/Librem_5/HiFi.conf [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+		cset "name='DAC L/R Swap Switch' off"
+		cset "name='DAC Monomix Switch' off"
+		cset "name='SPKOUTR PGA' Mixer"
+	]
+
+	Value {
+		PlaybackChannels "2"
+		PlaybackPriority "100"
+		PlaybackVolume "name='Speaker Volume'"
+		PlaybackSwitch "name='Speaker Switch'"
```
Comment (perexg) [inline ucm2/NXP/iMX8/Librem_5/HiFi.conf:57]: Missing 'PlaybackMixerElem "Speaker"` value for PA/PW to enable hardware volume control.

File: ucm2/NXP/iMX8/Librem_5/HiFi.conf [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+
+	DisableSequence [
+		cset "name='Headphone Switch' off off"
+		cset "name='HPOUTL PGA' 0 mute"
+		cset "name='HPOUTR PGA' 0 mute"
+	]
+
+	Value {
+		PlaybackPriority "1000"
+		PlaybackChannels "2"
```
Comment (perexg) [inline ucm2/NXP/iMX8/Librem_5/HiFi.conf:117]: 2 channels are the default. Remove this line.

File: ucm2/NXP/iMX8/Librem_5/HiFi.conf [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+		cset "name='Capture Switch' off,off"
+		cset "name='MIXINR PGA Switch' off,off"
+		cset "name='Input Mode' Digital"
+	]
+
+	Value {
+		CapturePriority "100"
+		CaptureChannels "2"
+		CaptureVolume "name='Capture Volume'"
+		CapturePCM "hw:${CardId},0"
```
Comment (perexg) [inline ucm2/NXP/iMX8/Librem_5/HiFi.conf:177]: First (zero) PCM device is the default. Use the shorter `CapturePCM "hw:${CardId}"`

**Review by perexg — CHANGES_REQUESTED** [review]
Please, resolve my objections for all devices (MixerElem, 2 channels, shorter PCM devices).

File: ucm2/NXP/iMX8/Librem_5/HiFi.conf [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+		cset "name='DAC L/R Swap Switch' off"
+		cset "name='DAC Monomix Switch' off"
+		cset "name='SPKOUTR PGA' Mixer"
+	]
+
+	Value {
+		PlaybackPriority "100"
+		PlaybackVolume "name='Speaker Volume'"
+		PlaybackSwitch "name='Speaker Switch'"
+                PlaybackMixerElem "Handset"
```
Comment (perexg) [inline ucm2/NXP/iMX8/Librem_5/HiFi.conf:28]: It should be "Speaker" (see `amixer scontrols`).

File: ucm2/NXP/iMX8/Librem_5/HiFi.conf [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+	]
+
+	DisableSequence [
+		cset "name='ADC L/R Swap Switch' off"
+	]
+
+	Value {
+		CapturePriority "500"
+		CaptureChannels "2"
+		CaptureVolume "name='Digital Capture Volume'"
```
Comment (perexg) [inline ucm2/NXP/iMX8/Librem_5/HiFi.conf:112]: Missing `CaptureMixerElem "Digital"` here.

File: ucm2/NXP/iMX8/Librem_5/HiFi.conf [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+	DisableSequence [
+		cset "name='Capture Switch' off,off"
+		cset "name='MIXINR PGA Switch' off,off"
+		cset "name='Input Mode' Digital"
+	]
+
+	Value {
+		CapturePriority "100"
+		CaptureChannels "2"
+		CaptureVolume "name='Capture Volume'"
```
Comment (perexg) [inline ucm2/NXP/iMX8/Librem_5/HiFi.conf:139]: Missing `CaptureMixerElem "Capture"` here.

File: ucm2/NXP/iMX8/Librem_5/HiFi.conf [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+
+	DisableSequence [
+		cset "name='Capture Switch' off,off"
+		cset "name='MIXINR PGA Switch' off,off"
+		cset "name='Input Mode' Digital"
+	]
+
+	Value {
+		CapturePriority "100"
+		CaptureChannels "2"
```
Comment (perexg) [inline ucm2/NXP/iMX8/Librem_5/HiFi.conf:138]: Remove CaptureChannels 2 (the default).

**Review by perexg — APPROVED** [review]
The verb should also call DisableSequence for all devices when enabled like: 

https://github.com/alsa-project/alsa-ucm-conf/blob/master/ucm2/HDA/HiFi.conf#L3-L9

**Pull request closed without merging**